File: src/ui.c
#include "../include/ui.h"
#include "../include/peer_client.h"
#include "../include/protocol.h"
#include "../include/network.h"
#include <signal.h>

static volatile int running = 1;

void signal_handler(int sig __attribute__((unused))) {
    running = 0;
}

/* Inicia interface */
void ui_start(PeerState *state) {
    char command[256];
    
    signal(SIGINT, signal_handler);
    
    print_help();
    
    while (running) {
        print_prompt();
        
        if (fgets(command, sizeof(command), stdin) == NULL) {
            break;
        }
        
        command[strcspn(command, "\n")] = 0;
        
        if (strlen(command) == 0) {
            continue;
        }
        
        ui_process_command(state, command);
    }
    
    if (state->joined) {
        cmd_leave(state);
    }
}

/* Processa comando */
void ui_process_command(PeerState *state, const char *command) {
    char arg[256];
    
    if (strcmp(command, "join") == 0) {
        cmd_join(state);
    } else if (strcmp(command, "leave") == 0) {
        cmd_leave(state);
    } else if (strcmp(command, "show neighbors") == 0) {
        cmd_show_neighbors(state);
    } else if (sscanf(command, "release %d", &(int){0}) == 1) {
        int seqnum;
        sscanf(command, "release %d", &seqnum);
        cmd_release(state, seqnum);
    } else if (strcmp(command, "list identifiers") == 0) {
        cmd_list_identifiers(state);
    } else if (sscanf(command, "post %s", arg) == 1) {
        cmd_post(state, arg);
    } else if (sscanf(command, "search %s", arg) == 1) {
        cmd_search(state, arg);
    } else if (sscanf(command, "unpost %s", arg) == 1) {
        cmd_unpost(state, arg);
    } else if (strcmp(command, "exit") == 0) {
        cmd_exit(state);
        running = 0;
    } else if (strcmp(command, "help") == 0) {
        print_help();
    } else {
        printf("Comando desconhecido. Digite 'help' para ajuda.\n");
    }
}

/* Comando join */
void cmd_join(PeerState *state) {
    if (state->joined) {
        printf("Já está na rede.\n");
        return;
    }
    
    /* Regista no servidor */
    if (peer_client_register(state->server_ip, state->server_port,
                            state->tcp_port, &state->seqnumber) < 0) {
        printf("Erro ao registar no servidor de peers.\n");
        return;
    }
    
    printf("Registado com seqnumber %d\n", state->seqnumber);
    
    /* Obtém lista de peers */
    PeerList peer_list;
    if (peer_client_get_peers(state->server_ip, state->server_port, 
                             &peer_list) < 0) {
        printf("Erro ao obter lista de peers.\n");
        peer_client_unregister(state->server_ip, state->server_port, 
                              state->seqnumber);
        return;
    }
    
    printf("Obtidos %d peers da rede.\n", peer_list.count);
    
    /* Estabelece ligações */
    int connected = establish_connections(state, &peer_list);
    
    state->joined = 1;
    
    if (connected > 0) {
        printf("Ligado à rede com %d vizinhos.\n", connected);
    } else if (peer_list.count == 0) {
        printf("És o primeiro peer na rede.\n");
    } else {
        printf("Aviso: Não conseguiu ligar-se a nenhum peer.\n");
    }
}

/* Comando leave */
void cmd_leave(PeerState *state) {
    if (!state->joined) {
        printf("Não está na rede.\n");
        return;
    }
    
    /* Remove registo */
    peer_client_unregister(state->server_ip, state->server_port, 
                          state->seqnumber);
    
    /* Fecha todas ligações */
    for (int i = 0; i < state->neighbors.count; i++) {
        close_socket(state->neighbors.neighbors[i].socket_fd);
    }
    
    state->neighbors.count = 0;
    state->joined = 0;
    
    printf("Abandonou a rede.\n");
}

/* Comando show neighbors */
void cmd_show_neighbors(PeerState *state) {
    if (!state->joined) {
        printf("Não está na rede.\n");
        return;
    }
    
    int external = 0, internal = 0;
    
    printf("\nVizinhos Externos:\n");
    for (int i = 0; i < state->neighbors.count; i++) {
        Neighbor *n = &state->neighbors.neighbors[i];
        if (n->is_external) {
            printf("  %s:%d seqnumber=%d\n", 
                   n->info.ip, n->info.port, n->info.seqnumber);
            external++;
        }
    }
    if (external == 0) printf("  (nenhum)\n");
    
    printf("\nVizinhos Internos:\n");
    for (int i = 0; i < state->neighbors.count; i++) {
        Neighbor *n = &state->neighbors.neighbors[i];
        if (!n->is_external) {
            printf("  %s:%d seqnumber=%d\n", 
                   n->info.ip, n->info.port, n->info.seqnumber);
            internal++;
        }
    }
    if (internal == 0) printf("  (nenhum)\n");
    
    printf("\n");
}

/* Comando release */
void cmd_release(PeerState *state, int seqnumber) {
    if (!state->joined) {
        printf("Não está na rede.\n");
        return;
    }
    
    Neighbor *neighbor = find_neighbor_by_seqnumber(state, seqnumber);
    if (!neighbor) {
        printf("Vizinho com seqnumber %d não encontrado.\n", seqnumber);
        return;
    }
    
    if (neighbor->is_external) {
        printf("Apenas pode remover vizinhos internos.\n");
        return;
    }
    
    if (remove_neighbor(state, seqnumber) == 0) {
        printf("Ligação com peer %d removida.\n", seqnumber);
    }
}

/* Comando list identifiers */
void cmd_list_identifiers(PeerState *state) {
    printf("\nIdentificadores conhecidos:\n");
    
    if (state->identifiers.count == 0) {
        printf("  (nenhum)\n");
    } else {
        for (int i = 0; i < state->identifiers.count; i++) {
            printf("  %s\n", state->identifiers.identifiers[i]);
        }
    }
    
    printf("\n");
}

/* Comando post */
void cmd_post(PeerState *state, const char *identifier) {
    if (state->identifiers.count >= MAX_IDENTIFIERS) {
        printf("Lista de identificadores cheia.\n");
        return;
    }
    
    /* Verifica se já existe */
    for (int i = 0; i < state->identifiers.count; i++) {
        if (strcmp(state->identifiers.identifiers[i], identifier) == 0) {
            printf("Identificador já existe.\n");
            return;
        }
    }
    
    strcpy(state->identifiers.identifiers[state->identifiers.count], identifier);
    state->identifiers.count++;
    
    printf("Identificador '%s' adicionado.\n", identifier);
}

/* Comando search */
void cmd_search(PeerState *state, const char *identifier) {
    if (!state->joined) {
        printf("Não está na rede.\n");
        return;
    }
    
    /* Verifica se já tem */
    for (int i = 0; i < state->identifiers.count; i++) {
        if (strcmp(state->identifiers.identifiers[i], identifier) == 0) {
            printf("Identificador '%s' já conhecido.\n", identifier);
            return;
        }
    }
    
    /* Pergunta aos vizinhos */
    int found = 0;
    for (int i = 0; i < state->neighbors.count; i++) {
        Neighbor *n = &state->neighbors.neighbors[i];
        
        protocol_query_identifier(n->socket_fd, identifier, 
                                 state->max_hopcount);
        
        char response[256];
        if (tcp_receive(n->socket_fd, response, sizeof(response)) > 0) {
            if (strncmp(response, "FND", 3) == 0) {
                found = 1;
                break;
            }
        }
    }
    
    if (found) {
        strcpy(state->identifiers.identifiers[state->identifiers.count], 
               identifier);
        state->identifiers.count++;
        printf("Identificador '%s' encontrado e adicionado.\n", identifier);
    } else {
        printf("Identificador '%s' não encontrado na rede.\n", identifier);
    }
}

/* Comando unpost */
void cmd_unpost(PeerState *state, const char *identifier) {
    for (int i = 0; i < state->identifiers.count; i++) {
        if (strcmp(state->identifiers.identifiers[i], identifier) == 0) {
            for (int j = i; j < state->identifiers.count - 1; j++) {
                strcpy(state->identifiers.identifiers[j],
                       state->identifiers.identifiers[j + 1]);
            }
            state->identifiers.count--;
            printf("Identificador '%s' removido.\n", identifier);
            return;
        }
    }
    
    printf("Identificador '%s' não encontrado.\n", identifier);
}

/* Comando exit */
void cmd_exit(PeerState *state) {
    printf("A sair...\n");
    cmd_leave(state);
}

/* Imprime prompt */
void print_prompt() {
    printf("p2pnet> ");
    fflush(stdout);
}

/* Imprime ajuda */
void print_help() {
    printf("\n=== P2P Overlay Network ===\n");
    printf("Comandos disponíveis:\n");
    printf("  join                    - Entrar na rede\n");
    printf("  leave                   - Sair da rede\n");
    printf("  show neighbors          - Mostrar vizinhos\n");
    printf("  release <seqnumber>     - Remover vizinho interno\n");
    printf("  list identifiers        - Listar identificadores\n");
    printf("  post <id>               - Adicionar identificador\n");
    printf("  search <id>             - Pesquisar identificador\n");
    printf("  unpost <id>             - Remover identificador\n");
    printf("  exit                    - Sair da aplicação\n");
    printf("  help                    - Mostrar esta ajuda\n");
    printf("\n");
}

File: ui.py
"""
Interface de usuário e comandos - Versão melhorada com cores
"""

import os
import sys
from peer_client import peer_client_register, peer_client_unregister, peer_client_get_peers
from protocol import (
    establish_connections, remove_neighbor, count_external_neighbors,
    find_neighbor_by_seqnumber, protocol_query_identifier, tcp_receive
)
from network import close_socket


# Cores ANSI para terminal
class Colors:
    """Cores para terminal"""
    HEADER = '\033[95m'
    BLUE = '\033[94m'
    CYAN = '\033[96m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    GRAY = '\033[90m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'
    END = '\033[0m'


def clear_screen():
    """Limpa a tela"""
    os.system('cls' if os.name == 'nt' else 'clear')


def print_header():
    """Imprime cabeçalho"""
    print(f"\n{Colors.BOLD}{Colors.CYAN}")
    print("╔═══════════════════════════════════════════════════════════╗")
    print("║                                                           ║")
    print("║          P2P Overlay Network - Python Edition            ║")
    print("║                                                           ║")
    print("╚═══════════════════════════════════════════════════════════╝")
    print(f"{Colors.END}")


def print_status(state):
    """Imprime status atual do peer"""
    status = f"{Colors.GREEN}✓ ONLINE{Colors.END}" if state.joined else f"{Colors.RED}✗ OFFLINE{Colors.END}"
    seqnum = f"#{state.seqnumber}" if state.seqnumber else "-"
    neighbors = len(state.neighbors.neighbors) if state.joined else 0
    
    print(f"\n{Colors.BOLD}Status:{Colors.END}")
    print(f"  Conexão: {status}")
    print(f"  SeqNumber: {seqnum}")
    print(f"  Vizinhos: {neighbors}")
    print(f"  Servidor: {state.server_ip}:{state.server_port}")
    print()


def print_help():
    """Imprime ajuda com formatação melhorada"""
    print(f"\n{Colors.BOLD}{Colors.YELLOW}=== COMANDOS DISPONÍVEIS ==={Colors.END}\n")
    
    commands = [
        ("join", "Entrar na rede overlay"),
        ("leave", "Sair da rede"),
        ("show neighbors", "Listar todos os vizinhos (internos e externos)"),
        ("release <id>", "Desconectar de um vizinho interno"),
        ("list identifiers", "Listar identificadores locais"),
        ("post <id>", "Publicar um novo identificador"),
        ("search <id>", "Procurar um identificador na rede"),
        ("unpost <id>", "Remover um identificador local"),
        ("status", "Ver status do peer"),
        ("help", "Mostrar esta ajuda"),
        ("clear", "Limpar tela"),
        ("exit", "Sair da aplicação"),
    ]
    
    for cmd, desc in commands:
        print(f"  {Colors.CYAN}{cmd:<25}{Colors.END} {desc}")
    print()


def print_prompt(state):
    """Imprime prompt com informação"""
    status = "ONLINE" if state.joined else "OFFLINE"
    color = Colors.GREEN if state.joined else Colors.RED
    print(f"{color}[{status}]{Colors.END} {Colors.BOLD}p2pnet>{Colors.END} ", end="", flush=True)


def cmd_join(state):
    """Comando join - Entrar na rede"""
    if state.joined:
        print(f"{Colors.YELLOW}[!] Já está na rede.{Colors.END}")
        return
    
    print(f"{Colors.BLUE}[*] Registando no servidor...{Colors.END}")
    
    # Regista no servidor
    result, seqnumber = peer_client_register(state.server_ip, state.server_port, state.tcp_port)
    if result < 0:
        return
    
    state.seqnumber = seqnumber
    
    # Obtém lista de peers
    print(f"{Colors.BLUE}[*] Obtendo lista de peers...{Colors.END}")
    result, peer_list = peer_client_get_peers(state.server_ip, state.server_port)
    if result < 0:
        print(f"{Colors.RED}[✗] Erro ao obter lista de peers.{Colors.END}")
        peer_client_unregister(state.server_ip, state.server_port, state.seqnumber)
        return
    
    peer_count = peer_list.count if peer_list else 0
    print(f"{Colors.BLUE}[*] Encontrados {peer_count} peers na rede.{Colors.END}")
    
    # Estabelece ligações
    connected = establish_connections(state, peer_list) if peer_list else 0
    
    state.joined = True
    
    if connected > 0:
        print(f"{Colors.GREEN}[✓] Ligado à rede com {connected} vizinhos!{Colors.END}")
    elif peer_list and peer_list.count == 0:
        print(f"{Colors.YELLOW}[!] És o primeiro peer na rede.{Colors.END}")
    else:
        print(f"{Colors.YELLOW}[!] Registado mas sem vizinhos conectados.{Colors.END}")


def cmd_leave(state):
    """Comando leave - Sair da rede"""
    if not state.joined:
        print(f"{Colors.YELLOW}[!] Não está na rede.{Colors.END}")
        return
    
    print(f"{Colors.BLUE}[*] Desconectando...{Colors.END}")
    
    # Remove registo
    peer_client_unregister(state.server_ip, state.server_port, state.seqnumber)
    
    # Fecha todas ligações
    for neighbor in state.neighbors.neighbors:
        close_socket(neighbor.socket_fd)
    
    state.neighbors.neighbors.clear()
    state.neighbors.count = 0
    state.joined = False
    
    print(f"{Colors.GREEN}[✓] Desconectado da rede.{Colors.END}")


def cmd_show_neighbors(state):
    """Comando show neighbors - Mostrar vizinhos"""
    if not state.joined:
        print(f"{Colors.YELLOW}[!] Não está na rede.{Colors.END}")
        return
    
    if not state.neighbors.neighbors:
        print(f"{Colors.YELLOW}[!] Nenhum vizinho conectado.{Colors.END}")
        return
    
    external = [n for n in state.neighbors.neighbors if n.is_external]
    internal = [n for n in state.neighbors.neighbors if not n.is_external]
    
    print(f"\n{Colors.BOLD}{Colors.CYAN}Vizinhos Externos ({len(external)}):{Colors.END}")
    if external:
        for n in external:
            print(f"  {Colors.CYAN}→{Colors.END} {n.info.ip}:{n.info.port} "
                  f"{Colors.GRAY}(seq: {n.info.seqnumber}){Colors.END}")
    else:
        print(f"  {Colors.GRAY}(nenhum){Colors.END}")
    
    print(f"\n{Colors.BOLD}{Colors.GREEN}Vizinhos Internos ({len(internal)}):{Colors.END}")
    if internal:
        for n in internal:
            print(f"  {Colors.GREEN}←{Colors.END} {n.info.ip}:{n.info.port} "
                  f"{Colors.GRAY}(seq: {n.info.seqnumber}){Colors.END}")
    else:
        print(f"  {Colors.GRAY}(nenhum){Colors.END}")
    print()


def cmd_release(state, seqnumber):
    """Comando release - Remover vizinho interno"""
    if not state.joined:
        print(f"{Colors.YELLOW}[!] Não está na rede.{Colors.END}")
        return
    
    neighbor = find_neighbor_by_seqnumber(state, seqnumber)
    if not neighbor:
        print(f"{Colors.RED}[✗] Vizinho #{seqnumber} não encontrado.{Colors.END}")
        return
    
    if neighbor.is_external:
        print(f"{Colors.YELLOW}[!] Apenas pode remover vizinhos internos.{Colors.END}")
        return
    
    if remove_neighbor(state, seqnumber) == 0:
        print(f"{Colors.GREEN}[✓] Ligação com peer #{seqnumber} removida.{Colors.END}")
    else:
        print(f"{Colors.RED}[✗] Erro ao remover vizinho.{Colors.END}")


def cmd_list_identifiers(state):
    """Comando list identifiers - Listar identificadores"""
    print(f"\n{Colors.BOLD}{Colors.CYAN}Identificadores Locais ({state.identifiers.count}):{Colors.END}")
    
    if state.identifiers.count == 0:
        print(f"  {Colors.GRAY}(nenhum){Colors.END}")
    else:
        for i, identifier in enumerate(state.identifiers.identifiers, 1):
            print(f"  {Colors.CYAN}{i}.{Colors.END} {identifier}")
    print()


def cmd_post(state, identifier):
    """Comando post - Adicionar identificador"""
    if state.identifiers.add(identifier) == 0:
        print(f"{Colors.GREEN}[✓] Identificador '{identifier}' publicado.{Colors.END}")
    else:
        print(f"{Colors.RED}[✗] Erro: Limite atingido ou já existe.{Colors.END}")


def cmd_search(state, identifier):
    """Comando search - Pesquisar identificador"""
    if not state.joined:
        print(f"{Colors.YELLOW}[!] Não está na rede.{Colors.END}")
        return
    
    # Verifica se já tem
    if state.identifiers.contains(identifier):
        print(f"{Colors.YELLOW}[!] Identificador '{identifier}' já conhece.{Colors.END}")
        return
    
    print(f"{Colors.BLUE}[*] Procurando '{identifier}' na rede...{Colors.END}")
    
    # Pergunta aos vizinhos
    found = False
    for neighbor in state.neighbors.neighbors:
        protocol_query_identifier(neighbor.socket_fd, identifier, state.max_hopcount)
        
        response = tcp_receive(neighbor.socket_fd)
        if response and response.startswith("FND"):
            found = True
            break
    
    if found:
        state.identifiers.add(identifier)
        print(f"{Colors.GREEN}[✓] Identificador '{identifier}' encontrado!{Colors.END}")
    else:
        print(f"{Colors.YELLOW}[!] Identificador '{identifier}' não encontrado.{Colors.END}")


def cmd_unpost(state, identifier):
    """Comando unpost - Remover identificador"""
    if state.identifiers.remove(identifier) == 0:
        print(f"{Colors.GREEN}[✓] Identificador '{identifier}' removido.{Colors.END}")
    else:
        print(f"{Colors.RED}[✗] Identificador não encontrado.{Colors.END}")


def cmd_status(state):
    """Comando status - Mostrar status"""
    print_status(state)


def cmd_exit(state):
    """Comando exit - Sair"""
    print(f"\n{Colors.BLUE}[*] Encerrando...{Colors.END}")
    if state.joined:
        cmd_leave(state)


def process_command(state, command):
    """Processa comando da UI"""
    command = command.strip()
    
    if not command:
        return True
    
    if command == "join":
        cmd_join(state)
    elif command == "leave":
        cmd_leave(state)
    elif command == "show neighbors":
        cmd_show_neighbors(state)
    elif command.startswith("release "):
        try:
            seqnumber = int(command.split()[1])
            cmd_release(state, seqnumber)
        except (ValueError, IndexError):
            print(f"{Colors.YELLOW}Uso: release <seqnumber>{Colors.END}")
    elif command == "list identifiers":
        cmd_list_identifiers(state)
    elif command.startswith("post "):
        identifier = command[5:].strip()
        if identifier:
            cmd_post(state, identifier)
        else:
            print(f"{Colors.YELLOW}Uso: post <identificador>{Colors.END}")
    elif command.startswith("search "):
        identifier = command[7:].strip()
        if identifier:
            cmd_search(state, identifier)
        else:
            print(f"{Colors.YELLOW}Uso: search <identificador>{Colors.END}")
    elif command.startswith("unpost "):
        identifier = command[7:].strip()
        if identifier:
            cmd_unpost(state, identifier)
        else:
            print(f"{Colors.YELLOW}Uso: unpost <identificador>{Colors.END}")
    elif command == "status":
        cmd_status(state)
    elif command == "help":
        print_help()
    elif command == "clear":
        clear_screen()
        print_header()
    elif command == "exit":
        cmd_exit(state)
        return False
    else:
        print(f"{Colors.RED}[✗] Comando desconhecido. Digite 'help' para ajuda.{Colors.END}")
    
    return True


def ui_start(state):
    """Inicia interface de usuário"""
    print_header()
    print_status(state)
    print_help()
    
    running = True
    while running:
        print_prompt(state)
        
        try:
            command = input()
            running = process_command(state, command)
        except KeyboardInterrupt:
            print(f"\n{Colors.BLUE}[*] Interrompido pelo utilizador...{Colors.END}")
            if state.joined:
                cmd_leave(state)
            break
        except EOFError:
            break
    
    if state.joined:
        cmd_leave(state)
    
    print(f"\n{Colors.CYAN}Obrigado por usar P2P Overlay Network!{Colors.END}\n")
